File: apps/web/src/features/checkout/api/mappers.ts
/**
 * Data transformation functions for checkout/payment API.
 */

import type {
  ConfirmPaymentInput,
  ConfirmPaymentResult,
  CreatePaymentOrderResult,
} from '../model';

export function toCreatePaymentOrderBody(bookingId: string) {
  return { bookingId };
}

export function toConfirmPaymentBody(input: ConfirmPaymentInput) {
  return {
    paymentKey: input.paymentKey,
    orderId: input.orderId,
    amount: input.amount,
  };
}

export function mapCreatePaymentOrderResult(data: unknown): CreatePaymentOrderResult | null {
  if (typeof data !== 'object' || data === null || !('orderId' in data) || !('amountKrw' in data)) {
    return null;
  }

  return data as CreatePaymentOrderResult;
}

export function mapConfirmPaymentResult(data: unknown): ConfirmPaymentResult | null {
  if (typeof data !== 'object' || data === null || !('status' in data)) {
    return null;
  }

  return data as ConfirmPaymentResult;
}

File: apps/backend-python/src/housing_platform/payments/mappers.py
import uuid

from housing_platform.db.models import Payments
from housing_platform.payments.schemas import (
    ConfirmPaymentResult,
    CreatePaymentOrderResult,
    PaymentStatus,
)


def map_create_payment_order_result(
    *,
    payment_id: uuid.UUID,
    order_id: uuid.UUID,
    booking_id: uuid.UUID,
    amount_krw: int,
    order_name: str,
) -> CreatePaymentOrderResult:
    return CreatePaymentOrderResult(
        payment_id=payment_id,
        order_id=order_id,
        booking_id=booking_id,
        amount_krw=amount_krw,
        order_name=order_name,
    )


def map_confirm_payment_result(
    payment: Payments,
    *,
    booking_id: uuid.UUID | None = None,
) -> ConfirmPaymentResult:
    return ConfirmPaymentResult(
        payment_id=payment.id,
        order_id=payment.order_id,
        booking_id=booking_id if booking_id is not None else payment.booking_id,
        status=PaymentStatus(payment.status),
    )
